src/event.cc

```diff
@@ -174,12 +165,14 @@ EventScheduler::cancel(EVH * func, void *arg)
     ev_entry *event;
 
     for (E = &tasks; (event = *E) != NULL; E = &(*E)->next) {
-        if (event->func != func)
-            continue;
 
-        if (arg && event->arg != arg)
+        const auto *dialer = dynamic_cast<EventDialer*>(event->call->getDialer());
+        assert(dialer);
+        if (!dialer->matches(func,arg))
             continue;
 
+        event->call->cancel("EventScheduler::cancel");
+
```

Comment:
This is so we can debug any issues with the event queue going away early during shutdown.

Reply:
> This is so we can debug any issues with the event queue going away early during shutdown.

FWIW, I do not see how cancelling the call here can help with that. AFAICT, both the event object and the call object in question will be gone by the time the current call stack unwinds a bit towards the event loop. Their state will be inaccessible in any triage. If you are after _debugging_ artifacts, then add debugs() to the cancel() method instead. That debugging can report useful information such as whether the event(s) were (not) found, and debugs() is much cheaper for non-debugging Squids than calling the cancel method.

One more reason to avoid this call cancellation is that it will confuse some developers into thinking that this EventScheduler::cancel() method always cancels the corresponding event notification. There were many bugs related to that confusion in the past.

Reply:
What I mean by debugging is that I am expecting the cache.log lines added by cancel() to show up if the Call has not been unwound like you think they should have been.

Reply:
> I am expecting the cache.log lines added by cancel() to show up if the Call has not been unwound like you think they should have been.

To avoid misunderstanding, I was talking about stack unwinding, not Call firing. Stack unwinding _will_ happen.

As for debugging I still support adding it! Here is a rough untested sketch:

```C++
debugs(41, 5, *event);
```

or, if you really do not want to add _proper_ event printing:

```C++
debugs(41, 5, event->call->id);
```

debugs() will automatically add the helpful "cancel" prefix, of course.

Reply:
Done in 43a6935